### PR TITLE
HDDS-5660. Separate container caches for Open and Closed containers.

### PR DESCRIPTION
**Pull request by mukul1987**
## What changes were proposed in this pull request?

It has been seen in deployments that with large number of containers on datanodes. we have seen cache misses on the rocksDB cache.

This patch proposes
a) to have separate caches for open and closed container
b) rocksDB containers in open container cache are never evicted.
c) Open containers are moved from open to close container cache it they are closed/quasi closed.
d) Closed container are cached in LRU fashion as they are cached right now

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5660

## How was this patch tested?
Added a new unit test

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/ContainerCache.java [hunk 1]
```diff
@@ -227,9 +246,68 @@ private boolean cleanupDb(ReferenceCountedDB db) {
...
+      Preconditions.checkNotNull(db, "Container not found in open cache:" +
+          containerPath);
+      closeContainerCache.putIfAbsent(containerPath, db);
+      metrics.incNumOpenContainerRemoves();
     } finally {
       lock.unlock();
     }
   }
+
+  public boolean isContainerOpen(State state) {
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/ContainerCache.java:269]: Would a better separation between the container states be `isContainerWritable`? Open is becoming an over loaded term.

Comment (mukul1987) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/ContainerCache.java:269]: ok will do that,

Comment (mukul1987) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/ContainerCache.java]: done in the followup patch

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/ContainerCache.java [hunk 2]
```diff
@@ -219,17 +238,80 @@ private boolean cleanupDb(ReferenceCountedDB db) {
...
+
+  private boolean isContainerWriteable(State state) {
+    switch (state) {
+    case CLOSED:
+    case QUASI_CLOSED:
+      return false;
+    case CLOSING:
+    case OPEN:
+    case UNHEALTHY:
+      return true;
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/ContainerCache.java:281]: Unhealthy containers are writeable?

**Comment by adoroszlai** [comment]
@mukul1987 `testZReadKeyWithUnhealthyContainerReplica` (in `TestOzoneRpcClient ` and `TestOzoneRpcClientWithRatis`) seems to be failing consistently.

**Pull request closed without merging**